src/module.cpp

```diff
@@ -1925,6 +1944,45 @@ bool Module::writeDevStub(const char *fn) {
     return true;
 }
 
+bool Module::writeCPPStub(const char *outFileName) { return writeCPPStub(this, outFileName); }
+
+bool Module::writeCPPStub(Module *module, const char *outFileName) {
+    // Get a file descriptor corresponding to where we want the output to
+    // go.  If we open it, it'll be closed by the llvm::raw_fd_ostream
+    // destructor.
+    //
+    int fd;
+    int flags = O_CREAT | O_WRONLY | O_TRUNC;
+
+    if (!outFileName) {
+        return false;
+    } else if (!strcmp("-", outFileName)) {
+        fd = 1;
+    } else {
+#ifdef ISPC_HOST_IS_WINDOWS
+        fd = _open(outFileName, flags, 0644);
+#else
+        fd = open(outFileName, flags, 0644);
+#endif // ISPC_HOST_IS_WINDOWS
+        if (fd == -1) {
+            perror(outFileName);
+            return false;
+        }
+    }
+
+    // The CPP stream should have been initialized: print and clean it up.
+    if (module->bufferCPP) {
+        llvm::raw_fd_ostream fos(fd, (fd != 1), false);
+        fos << module->bufferCPP->str;
+        module->bufferCPP.reset();
+    } else {
+        perror("Unexpected empty/null `CPPStream`");
+        return false;
+    }
```

Comment:
Would this qualify as a validity assertion, or is a call to `assert` preferred so that the check is removed under `Release`?

Reply:
I think it's better to have as an assert, as if nullptr comes here, it's really an internal error, which should never happen.

Note that `false` returned in multi-target compilation is currently ignored. This is probably need to be fixed, as well.

Reply:
Nice catch. Should be resolved now.

Reply:
> Nice catch. Should be resolved now.

Nevermind, there are some regressions.